packages/sources/s3reader/src/transport/csv.ts

```diff
@@ -0,0 +1,131 @@
+    await sleep(context.adapterSettings.BACKGROUND_EXECUTE_MS)
+  }
+
+  async handleRequest(param: RequestParams) {
+    let response: AdapterResponse<BaseEndpointTypes['Response']>
+    try {
+      response = await this._handleRequest(param)
+    } catch (e) {
+      const errorMessage = e instanceof Error ? e.message : 'Unknown error occurred'
+      logger.error(e, errorMessage)
+      response = {
+        statusCode: 502,
+        errorMessage,
+        timestamps: {
+          providerDataRequestedUnixMs: 0,
+          providerDataReceivedUnixMs: 0,
+          providerIndicatedTimeUnixMs: undefined,
+        },
+      }
+    }
+    await this.responseCache.write(this.name, [{ params: param, response }])
+  }
+
+  async _handleRequest(param: RequestParams): Promise<AdapterResponse<TransportTypes['Response']>> {
+    const { bucket, key, headerRow, resultColumn, matcherColumn, matcherValue } = param
+    const providerDataRequestedUnixMs = Date.now()
+    const csvFileAsStr = await getFileFromS3(this.s3Client, bucket, key)
+    const answer = this.parseCSV(csvFileAsStr, headerRow, matcherColumn, matcherValue, resultColumn)
+
+    return {
+      result: answer,
+      data: {
+        result: answer,
+      },
+      statusCode: 200,
+      timestamps: {
+        providerDataRequestedUnixMs,
+        providerDataReceivedUnixMs: Date.now(),
+        providerIndicatedTimeUnixMs: undefined,
+      },
+    }
+  }
+
+  // csvFileAsStr: CSV file as string as received from S3
+  // headerRow: 1-indexed row of the CSV file to use as the header row
+  // matcherColumn: field to match with `matcherValue` to find the answer row
+  // matcherValue: value of field `matcherColumn` used to find the answer row
+  // resultColumn: header field containing the answer in matcher row
+  parseCSV(
+    csvFileAsStr: string,
+    headerRow: number,
+    matcherColumn: string,
+    matcherValue: string,
+    resultColumn: string,
+  ): number {
+    // from_line is 1-indexed
+    // columns: true sets first line as object fields rather than 2d arrays
+    const parsed: Record<string, string>[] = parse(csvFileAsStr, {
+      columns: true,
+      from_line: headerRow,
+    })
+
+    const parsedHeaderRow = parsed[0]
+    // validate CSV contains headers matcherColumn and resultColumn
+    if (!(matcherColumn in parsedHeaderRow)) {
+      throw new Error(`CSV file does not contain column header ${matcherColumn}`)
+    }
+    if (!(resultColumn in parsedHeaderRow)) {
+      throw new Error(`CSV file does not contain column header ${resultColumn}`)
+    }
+
+    // find correct row using matcher
+    const matchingRows = parsed.filter((row) => row[matcherColumn] == matcherValue)
+    if (matchingRows.length === 0) {
+      throw new Error(`CSV file does not contain row where ${matcherColumn} == ${matcherValue}`)
+    }
+    if (matchingRows.length > 1) {
+      throw new Error(`CSV file contains multiple rows where ${matcherColumn} == ${matcherValue}`)
```

Comment:
Can we add a test, with a different csv to trigger this error?

Reply:
Added in https://github.com/smartcontractkit/external-adapters-js/pull/3313/commits/d06074b1a5b4d875daa0238bab4b2573d4247297#diff-a2f9928e24af6aedfb27b418061e6990f5a0141eb9c7cc0ffc2f1046abe1790cR251